STORM_JSTORM_FEATURE_DIFF.md

```diff
@@ -1,30 +1,30 @@
+Topology Master | Heartbeat server is similar for scaling.  Change is under review. | New system bolt "topology master" was added, which is responsible for collecting task heartbeat info of all tasks and reporting the info to nimbus. Besides task heartbeat info, it also can be used to dispatch control messages within the topology. Topology master significantly reduces the amout of read/write to ZooKeeper.	Before this change, ZooKeeper was the bottleneck for deploying big clusters and topologies.
+Backpressure | Store backpressure status in ZooKeeper, which can trigger the source spout to start "flow control", meaning it will stop sending tuples. | <ol><li>Implement backpressure using "topology master" (TM). TM is responsible for processing the trigger message and sending the flow control request to relevant spouts.  "flow control" in JStorm doesn't complete stop the spout from emitting tuples, but instead just slows down the tuple sending.<li>User can update the configuration of backpressure dynamically without restarting topology, e.g. enable/disable backpressue, high/low watermark, etc. | These two solutions are totally different, we need to evaluate which one is better. |
+Monitoring of task execute thread | N/A | Monitors the status of the execute thread of tasks. It is effective to find the slow bolt in a topology, and potentially uncovers deadlock as well. |
+Message processing | Configurable multi-receiving threads of worker (but likely to change to have deserialization happen in Netty) | <ol><li>Add receiving and transferring queue/thread for each task to make deserialization and serialization asynchronously<li>Remove receiving and transferring thread on worker level to avoid unnecessary locks and to shorten the message processing phase | Low risk for merging to JStorm (Both implementations are similar) |
+Batch tuples | Batching in DisruptorQueue | Do batch before sending tuple to transfer queue and support for adjusting the batch size dynamically according to samples of actual batch size sent out for past intervals. | Should evaluate both implementations, and see which is better, or whether we can move some of the dynamic batching logic into the disruptor level |
+Grouping | Load aware balancing shuffle grouping | <ol><li>Has a "localfirst" grouping that causes tuples to be sent to the tasks in the same worker by default. But if the load of all local tasks is high, the tuples will be sent out to remote tasks.<li>Improve localOrShuffle grouping that tuples are only sent to the tasks in the same worker or in same node.	The shuffle of JStorm checks the load of network connection of local worker. (QUESTION: this item isn't clear.)| The risk for merging of checking remote load which was implemented in "Load aware balancing shuffle grouping" is low.
```

Comment:
There is a localOrShuffle grouping where in the current version of storm if there is a bolt that it could send it to in the current worker process it will send the tuples there.  If there is not one, it will do round robin between all of the instances of that bolt no matter where they are.   JStorm has extended that from just within the same JVM but to be on the same host, even if they are in different workers.